File: pgpainless-core/src/main/java/org/pgpainless/key/generation/KeySpec.java
// SPDX-FileCopyrightText: 2018 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import org.bouncycastle.openpgp.PGPSignatureSubpacketVector;
import org.pgpainless.algorithm.KeyFlag;
import org.pgpainless.key.generation.type.KeyType;
import org.pgpainless.signature.subpackets.SignatureSubpackets;
import org.pgpainless.signature.subpackets.SignatureSubpacketsHelper;

import java.util.Date;

public class KeySpec {

    private final KeyType keyType;
    private final SignatureSubpackets subpacketGenerator;
    private final boolean inheritedSubPackets;
    private final Date keyCreationDate;

    KeySpec(@Nonnull KeyType type,
            @Nonnull SignatureSubpackets subpacketGenerator,
            boolean inheritedSubPackets,
            @Nullable Date keyCreationDate) {
        this.keyType = type;
        this.subpacketGenerator = subpacketGenerator;
        this.inheritedSubPackets = inheritedSubPackets;
        this.keyCreationDate = keyCreationDate;
    }

    @Nonnull
    public KeyType getKeyType() {
        return keyType;
    }

    @Nonnull
    public PGPSignatureSubpacketVector getSubpackets() {
        return SignatureSubpacketsHelper.toVector(subpacketGenerator);
    }

    @Nonnull
    public SignatureSubpackets getSubpacketGenerator() {
        return subpacketGenerator;
    }

    boolean isInheritedSubPackets() {
        return inheritedSubPackets;
    }

    @Nullable
    public Date getKeyCreationDate() {
        return keyCreationDate;
    }

    public static KeySpecBuilder getBuilder(KeyType type, KeyFlag flag, KeyFlag... flags) {
        return new KeySpecBuilder(type, flag, flags);
    }
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/key/generation/KeySpec.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation

import java.util.*
import org.pgpainless.algorithm.AEADCipherMode
import org.pgpainless.algorithm.CompressionAlgorithm
import org.pgpainless.algorithm.Feature
import org.pgpainless.algorithm.HashAlgorithm
import org.pgpainless.algorithm.KeyFlag
import org.pgpainless.algorithm.SymmetricKeyAlgorithm
import org.pgpainless.key.generation.type.KeyType

data class KeySpec(
    val keyType: KeyType,
    val keyFlags: List<KeyFlag>,
    val preferredCompressionAlgorithmsOverride: Set<CompressionAlgorithm>?,
    val preferredHashAlgorithmsOverride: Set<HashAlgorithm>?,
    val preferredSymmetricAlgorithmsOverride: Set<SymmetricKeyAlgorithm>?,
    val preferredAEADAlgorithmsOverride: Set<AEADCipherMode>?,
    val featuresOverride: Set<Feature>?,
    val keyCreationDate: Date?
) {

    companion object {
        @JvmStatic
        fun getBuilder(type: KeyType, vararg flags: KeyFlag) = KeySpecBuilder(type, *flags)
    }
}
